File: backend/GameApp.Domain/Entities/Scenes/TradeScene.cs
using GameApp.Domain.Entities.Items;
using GameApp.Domain.Enumerates;
using GameApp.Domain.ValueObjects.Scenes;

namespace GameApp.Domain.Entities.Scenes;

// Class for scene where you trade with a merchant
public class TradeScene : Scene
{
    private readonly int MerchantMoneyToSpent;
    private readonly List<Item> MerchantItemsOffer;
    private readonly int ProfitMerchantMargin;

    public TradeScene(
        SceneName name,
        SceneDescription description,
        Biome biome,
        int merchantMoneyToSpent,
        List<Item> merchantItemsOffer,
        int profitMerchantMargin
    ) : base(name, description, biome)
    {
        MerchantMoneyToSpent = merchantMoneyToSpent;
        MerchantItemsOffer = merchantItemsOffer;
        ProfitMerchantMargin = profitMerchantMargin;
    }

    public TradeScene(
        Guid id,
        SceneName name,
        SceneDescription description,
        Biome biome,
        int merchantMoneyToSpent,
        List<Item> merchantItemsOffer,
        int profitMerchantMargin
    ) : base(id, name, description, biome)
    {
        MerchantMoneyToSpent = merchantMoneyToSpent;
        MerchantItemsOffer = merchantItemsOffer;
        ProfitMerchantMargin = profitMerchantMargin;
    }

    public int GetMerchantMoneyToSpent() => MerchantMoneyToSpent;
    public List<Item> GetMerchantItemsOffer() => MerchantItemsOffer;
    public int GetProfitMerchantMargin() => ProfitMerchantMargin;

    public TradeScene SetSceneName(SceneName newName) =>
        new TradeScene(newName, GetDescription(), GetBiome(), MerchantMoneyToSpent, MerchantItemsOffer, ProfitMerchantMargin);

    public TradeScene SetSceneDescription(SceneDescription newDescription) =>
        new TradeScene(GetName(), newDescription, GetBiome(), MerchantMoneyToSpent, MerchantItemsOffer, ProfitMerchantMargin);

    public TradeScene SetBiome(Biome newBiome) =>
        new TradeScene(GetName(), GetDescription(), newBiome, MerchantMoneyToSpent, MerchantItemsOffer, ProfitMerchantMargin);

    public TradeScene SetMerchantMoneyToSpent(int newMerchantMoneyToSpent) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), newMerchantMoneyToSpent, MerchantItemsOffer, ProfitMerchantMargin);

    public TradeScene SetMerchantItemsOffer(List<Item> newMerchantItems) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), MerchantMoneyToSpent, newMerchantItems, ProfitMerchantMargin);

    public TradeScene SetProfitMerchantMargin(int newMargin) =>
        new TradeScene(GetName(), GetDescription(), GetBiome(), MerchantMoneyToSpent, MerchantItemsOffer, newMargin);

    public override string ToString()
    {
        string merchantItems = string.Join(", ", MerchantItemsOffer.Select(i => i.ToString()));

        return $"TradeScene({GetGuid()}): {GetName()} - {GetBiome()}\n" +
               $"Merchant Money To Spend: {MerchantMoneyToSpent}\n" +
               $"Merchant Profit Margin: {ProfitMerchantMargin}\n" +
               $"Merchant Items Offer: [{merchantItems}]";
    }
}

File: frontend/gameapp-frontend/src/domain/entities/scenes/trade-scene.ts
import { Scene } from "./scene";
import { SceneName } from "../../value-objects/scenes/scene-name";
import { SceneDescription } from "../../value-objects/scenes/scene-description";
import { Biome } from "../../enumerates/biome";
import { Item } from "../items/item";

export class TradeScene extends Scene {
    protected _merchantMoneyToSpent: number;
    protected _merchantItemsOffer: Item[];
    protected _profitMerchantMargin: number;

    constructor(
        name: SceneName,
        description: SceneDescription,
        biome: Biome,
        merchantMoneyToSpent: number,
        merchantItemsOffer: Item[],
        profitMerchantMargin: number,
        id?: string
    ) {
        super(name, description, biome, id);
        this._merchantMoneyToSpent = merchantMoneyToSpent;
        this._merchantItemsOffer = [...merchantItemsOffer];
        this._profitMerchantMargin = profitMerchantMargin;
    }

    get merchantMoneyToSpent(): number { return this._merchantMoneyToSpent; }
    get merchantItemsOffer(): Item[] { return [...this._merchantItemsOffer]; }
    get profitMerchantMargin(): number { return this._profitMerchantMargin; }

    setSceneName(newName: SceneName) {
        return new TradeScene(newName, this._description, this._biome, this._merchantMoneyToSpent, this._merchantItemsOffer, this._profitMerchantMargin, this._id);
    }
    setSceneDescription(newDescription: SceneDescription) {
        return new TradeScene(this._name, newDescription, this._biome, this._merchantMoneyToSpent, this._merchantItemsOffer, this._profitMerchantMargin, this._id);
    }
    setBiome(newBiome: Biome) {
        return new TradeScene(this._name, this._description, newBiome, this._merchantMoneyToSpent, this._merchantItemsOffer, this._profitMerchantMargin, this._id);
    }
    setMerchantMoneyToSpent(newMoney: number) {
        return new TradeScene(this._name, this._description, this._biome, newMoney, this._merchantItemsOffer, this._profitMerchantMargin, this._id);
    }
    setMerchantItemsOffer(newItems: Item[]) {
        return new TradeScene(this._name, this._description, this._biome, this._merchantMoneyToSpent, newItems, this._profitMerchantMargin, this._id);
    }
    setProfitMerchantMargin(newMargin: number) {
        return new TradeScene(this._name, this._description, this._biome, this._merchantMoneyToSpent, this._merchantItemsOffer, newMargin, this._id);
    }

    toString(): string {
        const merchantItems = this._merchantItemsOffer.map(i => i.toString()).join(", ");
        return `TradeScene(${this._id}): ${this._name} - ${this._biome}\n` +
               `Merchant Money To Spent: ${this._merchantMoneyToSpent}\n` +
               `Profit Merchant Margin: ${this._profitMerchantMargin}\n` +
               `Merchant Items: [${merchantItems}]`;
    }
}
